gnovm/pkg/packages/pkglist.go

```diff
@@ -0,0 +1,110 @@
+package packages
+
+import (
+	"errors"
+	"fmt"
+)
+
+type (
+	PkgList       []*Package
+	SortedPkgList []*Package
+)
+
+func (pl PkgList) Get(pkgPath string) *Package {
+	for _, p := range pl {
+		if p.ImportPath == pkgPath {
+			return p
+		}
+	}
+	return nil
+}
+
+var ErrPackageNotFound = errors.New("package not found")
+
+func (pl PkgList) GetByDir(dir string) *Package {
+	for _, p := range pl {
+		if p.Dir == dir {
+			return p
+		}
+	}
+	return nil
+}
+
+// sortPkgs sorts the given packages by their dependencies.
+func (pl PkgList) Sort() (SortedPkgList, error) {
+	visited := make(map[string]bool)
+	onStack := make(map[string]bool)
```

Comment:
maybe use map[string]struct{} if you just care about presence

Reply:
it's an interesting (although bikesheding) debate actually, I'm usually more leaned towards using `struct{}` as map value for sets since the `false` state in bool map is redundant with "having no value at this key" and it saves memory. But it makes the code less readable and "early optimizations is the root of all evil". The guys from the "The Go Programming Language" book say:
``` 
Some Go programmers use it instead of bool as the value type of a map that represents a set, to emphasize that only the keys are significant, but the space saving is marginal and the syntax more cumbersome, so we generally avoid it.
```
I think after diving into this I'm more leaned towards using bool for ease of reading and writing and change if the code becomes a bottleneck.